File: src/shmem/runtime.cpp
// Copyright © Advanced Micro Devices, Inc. All rights reserved.
//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// Copyright © Advanced Micro Devices, Inc. All rights reserved.
// MIT License
//
// Shmem runtime APIs: module management, barriers, query functions,
// and GpuStates management (shared with init.cpp).

#include <cassert>
#include <vector>

#include "hip/hip_runtime_api.h"
#include "mori/shmem/internal.hpp"
#include "mori/shmem/shmem_api.hpp"
#include "mori/utils/hip_helper.hpp"
#include "mori/utils/mori_log.hpp"

namespace mori {
namespace shmem {

/* ---------------------------------------------------------------------------------------------- */
/*                                  JIT Module & GpuStates Management                            */
/* ---------------------------------------------------------------------------------------------- */

using GpuStatesAddrProvider = void* (*)();

namespace {
// Meyer's singleton: initialized on first use, avoiding static-init order fiasco
std::vector<GpuStatesAddrProvider>& GpuStatesProviders() {
  static std::vector<GpuStatesAddrProvider> instance;
  return instance;
}
}  // namespace

using BarrierLauncher = void (*)(hipStream_t);
static BarrierLauncher s_staticBarrierLauncher = nullptr;

void RegisterGpuStatesAddrProvider(GpuStatesAddrProvider provider) {
  GpuStatesProviders().push_back(provider);
}

void RegisterBarrierLauncher(BarrierLauncher launcher) { s_staticBarrierLauncher = launcher; }

size_t GetGpuStatesAddrProviderCount() { return GpuStatesProviders().size(); }

int LoadShmemModule(const char* hsaco_path) {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  ModuleStates& ms = states->moduleStates;

  if (ms.module != nullptr) return 0;
  hipError_t err = hipModuleLoad(&ms.module, hsaco_path);
  if (err != hipSuccess) {
    MORI_SHMEM_ERROR("Failed to load shmem module from {}: {}", hsaco_path, hipGetErrorString(err));
    return -1;
  }
  err = hipModuleGetGlobal(reinterpret_cast<hipDeviceptr_t*>(&ms.gpuStatesPtr), nullptr, ms.module,
                           "_ZN4mori5shmem15globalGpuStatesE");
  if (err != hipSuccess) {
    MORI_SHMEM_ERROR("globalGpuStates symbol not found in shmem module: {}",
                     hipGetErrorString(err));
    return -1;
  }
  err = hipModuleGetFunction(&ms.barrierFunc, ms.module, "mori_shmem_barrier_all_block");
  if (err != hipSuccess) {
    MORI_SHMEM_ERROR("mori_shmem_barrier_all_block not found in shmem module: {}",
                     hipGetErrorString(err));
    return -1;
  }
  MORI_SHMEM_TRACE("Loaded shmem JIT module: globalGpuStates={:p}, barrier={:p}",
                   (void*)ms.gpuStatesPtr, (void*)ms.barrierFunc);
  return 0;
}

void CopyGpuStatesToDevice(ShmemStates* states) {
  const GpuStates* gpuStates = &states->gpuStates;
  ModuleStates& ms = states->moduleStates;

  if (ms.gpuStatesPtr != nullptr) {
    MORI_SHMEM_TRACE("Copying GpuStates to JIT module globalGpuStates ({:p})",
                     (void*)ms.gpuStatesPtr);
    HIP_RUNTIME_CHECK(
        hipMemcpy(ms.gpuStatesPtr, gpuStates, sizeof(GpuStates), hipMemcpyHostToDevice));
  }

  for (auto& provider : GpuStatesProviders()) {
    void* staticAddr = provider();
    if (staticAddr != nullptr) {
      MORI_SHMEM_TRACE("Copying GpuStates to static globalGpuStates ({:p})", staticAddr);
      HIP_RUNTIME_CHECK(hipMemcpy(staticAddr, gpuStates, sizeof(GpuStates), hipMemcpyHostToDevice));
    }
  }

  MORI_SHMEM_TRACE("Successfully copied GpuStates to device (rank={}, worldSize={})",
                   gpuStates->rank, gpuStates->worldSize);
}

void FinalizeRuntime(ShmemStates* states) {
  ModuleStates& ms = states->moduleStates;
  if (ms.module != nullptr) {
    hipModuleUnload(ms.module);
    ms.module = nullptr;
    ms.gpuStatesPtr = nullptr;
    ms.barrierFunc = nullptr;
  }
  states->gpuStates = {};
}

/* ---------------------------------------------------------------------------------------------- */
/*                                      Module Initialization                                    */
/* ---------------------------------------------------------------------------------------------- */

int ShmemModuleInit(void* hipModule) {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  states->CheckStatusValid();

  hipModule_t module = static_cast<hipModule_t>(hipModule);
  GpuStates* moduleGlobalGpuStatesAddr = nullptr;

  hipError_t err = hipModuleGetGlobal(reinterpret_cast<hipDeviceptr_t*>(&moduleGlobalGpuStatesAddr),
                                      nullptr, module, "_ZN4mori5shmem15globalGpuStatesE");

  if (err != hipSuccess) {
    (void)hipGetLastError();
    MORI_SHMEM_TRACE("Module does not contain globalGpuStates symbol ({}), skipping init",
                     hipGetErrorString(err));
    return -1;
  }

  MORI_SHMEM_TRACE("Module globalGpuStates address: {:p} (JIT module address: {:p})",
                   (void*)moduleGlobalGpuStatesAddr, (void*)states->moduleStates.gpuStatesPtr);

  HIP_RUNTIME_CHECK(hipMemcpy(moduleGlobalGpuStatesAddr, &states->gpuStates, sizeof(GpuStates),
                              hipMemcpyHostToDevice));

  MORI_SHMEM_TRACE("Successfully initialized globalGpuStates in module (rank={}, worldSize={})",
                   states->gpuStates.rank, states->gpuStates.worldSize);

  return 0;
}

int CopyGpuStatesToSymbol(void* deviceSymbolAddr) {
  if (deviceSymbolAddr == nullptr) return -1;
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  HIP_RUNTIME_CHECK(
      hipMemcpy(deviceSymbolAddr, &states->gpuStates, sizeof(GpuStates), hipMemcpyHostToDevice));
  return 0;
}

/* ---------------------------------------------------------------------------------------------- */
/*                                      Query APIs                                               */
/* ---------------------------------------------------------------------------------------------- */

int ShmemMyPe() {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  return states->bootStates->rank;
}

int ShmemNPes() {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  return states->bootStates->worldSize;
}

int ShmemNumQpPerPe() {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  return states->rdmaStates->commContext->GetNumQpPerPe();
}

bool ShmemSdmaEnabled() {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  return states->rdmaStates->commContext->IsSdmaEnabled();
}

/* ---------------------------------------------------------------------------------------------- */
/*                                      Barrier APIs                                             */
/* ---------------------------------------------------------------------------------------------- */

void ShmemBarrierAll() {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  states->CheckStatusValid();

  MORI_SHMEM_TRACE("PE {} entering barrier", states->bootStates->rank);
  states->bootStates->bootNet->Barrier();
  MORI_SHMEM_TRACE("PE {} exiting barrier", states->bootStates->rank);
}

void ShmemBarrierOnStream(hipStream_t stream) {
  ShmemStates* states = ShmemStatesSingleton::GetInstance();
  states->CheckStatusValid();

  MORI_SHMEM_TRACE("PE {} launching device barrier on stream", states->bootStates->rank);

  if (states->moduleStates.barrierFunc != nullptr) {
    hipError_t err = hipModuleLaunchKernel(states->moduleStates.barrierFunc, 1, 1, 1, 1, 1, 1, 0,
                                           stream, nullptr, nullptr);
    assert(err == hipSuccess && "ShmemBarrierOnStream launch failed");
  } else if (s_staticBarrierLauncher != nullptr) {
    s_staticBarrierLauncher(stream);
  } else {
    MORI_SHMEM_ERROR(
        "ShmemBarrierOnStream: no barrier kernel available. "
        "Load JIT shmem module (Python) or include shmem.hpp (C++ hipcc).");
    assert(false);
  }
}

}  // namespace shmem
}  // namespace mori

File: python/mori/ir/flydsl/runtime.py
# Copyright © Advanced Micro Devices, Inc. All rights reserved.
#
# MIT License
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
"""
FlyDSL-specific runtime helpers for mori shmem integration.

  - ``get_bitcode_path()``  — returns the path to libmori_shmem_device.bc
  - ``shmem_module_init()`` — initializes ``globalGpuStates`` for a loaded
                              FlyDSL HIP module
"""

from mori.ir.bitcode import find_bitcode

_FLYDSL_COV = 6
_FLYDSL_ROCDL_ABI = 600


def _check_flydsl_rocdl_abi() -> None:
    """Verify FlyDSL still lowers ROCm code objects with ABI 600.

    Mori's FlyDSL device bitcode is compiled with code object version 6.  The
    matching FlyDSL `rocdl-attach-target` option is `abi=600`; if either side
    changes, linked kernels can compile but fail later at HIP module load time.
    """
    try:
        from flydsl.compiler.backends import get_backend

        backend = get_backend()
        fragments = backend.pipeline_fragments(compile_hints={})
    except Exception as exc:
        raise RuntimeError(
            "Unable to inspect FlyDSL ROCm backend ABI settings"
        ) from exc

    attach_fragments = [frag for frag in fragments if "rocdl-attach-target" in frag]
    if not attach_fragments:
        raise RuntimeError("FlyDSL ROCm pipeline has no rocdl-attach-target pass")

    expected = f"abi={_FLYDSL_ROCDL_ABI}"
    if not any(expected in frag for frag in attach_fragments):
        raise RuntimeError(
            "MORI FlyDSL bitcode is built with cov=6 and requires FlyDSL "
            f"rocdl-attach-target {expected}; got: {attach_fragments}"
        )


def get_bitcode_path() -> str:
    """Return the path to libmori_shmem_device.bc (compiled with cov=6 for FlyDSL ABI).

    Usage::

        from mori.ir.flydsl import get_bitcode_path
        bc = get_bitcode_path()
    """
    _check_flydsl_rocdl_abi()
    return find_bitcode(cov=_FLYDSL_COV)


def shmem_module_init(hip_module: int):
    """Initialize globalGpuStates in a FlyDSL-loaded HIP module."""
    import mori.shmem as ms

    return ms.shmem_module_init(hip_module)


def install_hook() -> None:
    """Compatibility no-op.

    Modern FlyDSL integration attaches ``shmem_module_init`` directly through
    ``link_extern(..., module_init_fn=...)`` when constructing the extern
    wrappers, so no global hook installation is required.
    """
    return None


def install_jit_hook() -> None:
    """Compatibility alias for :func:`install_hook`."""
    return install_hook()
